db/mysql.js

```diff
@@ -676,5 +677,90 @@ module.exports = function (log, error) {
     )
   }
 
+  // Execute a callback with an ordered list of unpublished events.
+  //
+  // This method will fetch the next batch of unpublished events from
+  // the db and pass them to the provided callback, which should return
+  // a promise.  If the callback completes successfully then the events
+  // will be marked as published; if it completes with an error then the
+  // events will remain unpublished and available for a subsequent fetch.
+  //
+  // If the callback succeeds in processing some, but not all, of the events
+  // then it should return the number of events successfully processed.
+  // We assume they're processed in order.
+  //
+  // The database enforces mutual exclusion of access to events in order to
+  // prevent publication of duplicates.  If another client is currently
+  // accessing the events, then the callback is not invoked and this method
+  // will reject with an EventQueueLockedError error object.
+
+  var GET_UNPUBLISHED_EVENTS = 'CALL getUnpublishedEvents_1()'
+  var ACK_PUBLISHED_EVENTS = 'CALL ackPublishedEvents_1(?)'
+
+  function EventQueueLockedError(message) {
+    this.message = message || 'event queue locked'
+    Error.captureStackTrace(this, EventQueueLockedError)
+  }
+  util.inherits(EventQueueLockedError, Error)
+
+  MySql.prototype.processUnpublishedEvents = function (callback, errback) {
+    // Internally we use a db-level lock for mutual exclusion.
+    // We have to call GET_LOCK() and RELEASE_LOCK() on the same
+    // db connection or things will get very confused.
+    return this.getConnection('MASTER')
+      .then(
+        function (connection) {
+          // We must manually release this connection when we're done.
+          return query(connection, GET_UNPUBLISHED_EVENTS, [])
+            .then(
+              function (results) {
+                // Did we successfully acquire the lock?
+                if (!results[0][0].lockAcquired) {
+                  throw new EventQueueLockedError()
+                }
+                // We're now holding a db-level lock that must be released.
+                // Luckily MySQL will clear it if our connection dies,
+                // and we kill the connection in the event of error.
+                var events = results[2]
+                return callback(events)
```

Comment:
nit - this is an unconventional use for a "callback" in node, so maybe we should name it something else. I'd use `fn` to keep it short.